lrauv_ignition_plugins/test/test_ahrs.cc

```diff
@@ -0,0 +1,527 @@
+     this->fixture->Server()->Run(blocking, iterations, !paused);
+   } while (this->simTime < simTimeLimit);
+ }
+
+ protected: void SteerInENU(const ignition::math::Quaterniond &_rotation)
+ {
+   this->poseRequest = ignition::math::Pose3d(
+       ignition::math::Vector3d::Zero, _rotation);
+ }
+
+ protected: ignition::math::Vector3d LastAngularVelocityMeasurement() const
+ {
+   std::lock_guard<std::mutex> lock(lastImuMessageMutex);
+   return ignition::msgs::Convert(this->lastImuMessage.angular_velocity());
+ }
+
+ protected: ignition::math::Vector3d LastLinearAccelerationMeasurement() const
+ {
+   std::lock_guard<std::mutex> lock(lastImuMessageMutex);
+   return ignition::msgs::Convert(this->lastImuMessage.linear_acceleration());
+ }
+
+ protected: ignition::math::Vector3d LastMagneticFieldMeasurement() const
+ {
+   std::lock_guard<std::mutex> lock(lastMagnetometerMessageMutex);
+   return ignition::msgs::Convert(this->lastMagnetometerMessage.field_tesla());
+ }
+
+ protected: ignition::math::Quaterniond LastOrientationEstimate() const
+ {
+   std::lock_guard<std::mutex> lock(lastImuMessageMutex);
+   return ignition::msgs::Convert(this->lastImuMessage.orientation());
+ }
+
+ private: void OnImuMessage(const ignition::msgs::IMU &_msg)
+ {
+   std::lock_guard<std::mutex> lock(lastImuMessageMutex);
+   this->lastImuMessage = _msg;
+ }
+
+ private: void OnMagnetometerMessage(const ignition::msgs::Magnetometer &_msg)
+ {
+   std::lock_guard<std::mutex> lock(lastMagnetometerMessageMutex);
+   this->lastMagnetometerMessage = _msg;
+ }
+
+ private: ignition::msgs::IMU lastImuMessage;
+
+ private: mutable std::mutex lastImuMessageMutex;
+
+ private: ignition::msgs::Magnetometer lastMagnetometerMessage;
+
+ private: mutable std::mutex lastMagnetometerMessageMutex;
+
+ private: std::optional<ignition::math::Pose3d> poseRequest;
+
+ private: std::unique_ptr<ignition::gazebo::TestFixture> fixture;
+
+ private: std::chrono::steady_clock::duration simTime{
+   std::chrono::steady_clock::duration::zero()};
+
+ private: double maxStepSize;
+
+ private: ignition::transport::Node node;
+};
+
+// Acceleration due to gravity
+static constexpr double g0{9.8};  // m/s^2
+// Earth magnetic field
+static constexpr double Be{6.0e-6};  // T
+static constexpr double Bn{2.3e-5};  // T
+static constexpr double Bu{-4.2e-5};  // T
+
+static constexpr double angularVelocityTolerance{1e-4};
+static constexpr double linearAccelerationTolerance{1e-4};
+static constexpr double magneticFieldTolerance{1e-8};
+static constexpr double orientationTolerance{1e-4};
+
+inline bool AreQuaternionsEqual(
+    const ignition::math::Quaterniond &_qa,
+    const ignition::math::Quaterniond &_qb,
+    const double _tolerance)
+{
+  // Use geodesic distance between quaternions
+  return std::acos(2. * std::pow(_qa.Dot(_qb), 2) - 1.) < _tolerance;
+}
+
+//////////////////////////////////////////////////
+TEST_F(AhrsTest, FrameConventionsAreCorrect)
```

Comment:
@braanan @arjo129 at first I tried to break this test into smaller ones, but I'm unable to instantiate `ignition::gazebo::TestFixture` more than once in the same process. It aborts with `CHECK failed: GeneratedDatabase()->Add(...)`, which appears to be coming from the shared library that bears `lrauv`
protobuf messages. I suspect the Ignition plugin system is attempting to reload it upon fixture construction (see [here](https://groups.google.com/g/protobuf/c/RZRCnOywdSk/m/s5QvifLCBgAJ?utm_medium=email&utm_source=footer)).

Reply:
Oh yeah I forgot to mention I think we keep each test isolated in a separate process as there was a bug in the way we are linking the messages. For whatever reason if we rely on external messages google protobuf will go :boom:. I have not been able to isolate exactly why this is the case.